### PR TITLE
Enabled 3LO loopback tests

### PR DESCRIPTION
**Pull request by ulisesL**
Hi Andy,

3LO loopback testing is complete - all tests are passing!

FYI: 
I found some bugs in the cache that were a little difficult to debug. In summary, the cache key was not working. Mainly because the regex exp did not take into account new lines. The credentials file provided in the tests consisted of multiple lines. In contrast, the file provided by GCP had a single line. (see further notes below)

TODO:
I tried to split the test file into multiple, but there are many technicalities that kept me from it. I think that should be a task of its own and did not want to make it part of this commit. 

Please let me know if you have any further questions. As always, I tried to make the code as readable as possible!

Thanks for reviewing,
Ulises

### CODE REVIEW

File: integration/fixtures/fake-client-secrets-3lo-loopback.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "installed": {
     "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
     "auth_uri": "https://accounts.google.com/o/oauth2/auth",
-    "client_email": "",
+    "client_email": "force-cache-diff",
```
Comment (ulisesL) [inline integration/fixtures/fake-client-secrets-3lo-loopback.json:5]: FYI: this is necessary when running commands that use the cache. For example, tests in the original 3LO that do not bypass the cache will create a new cache entry. When tests for the loopback version run without bypassing the cache will return the previously generated cache entry - the same cache entry generated by the out of band version. This is because the redirect_uri is completely removed when creating a cache entry. In other words, redirect_uris={A,B} == redirect_uris={B,A}.

File: util/cache.go [hunk 2]
```diff
@@ -132,7 +132,7 @@ func createKey(settings *Settings) CacheKey {
 	// Removing redirect_uri from credentials file. This allows for dynamic
 	// localhost ports created during 3LO loopback.
 	var credentialsJSON string = settings.CredentialsJSON
-	re := regexp.MustCompile("\"redirect_uris\":\\[(.*?)\\]")
+	re := regexp.MustCompile("\"redirect_uris\":([[:graph:]\\s]*?)\\]")
```
Comment (ulisesL) [inline util/cache.go:135]: This regex takes into account new lines.

File: integration/cli_test.go [hunk 3]
```diff
@@ -316,88 +337,166 @@ func Test3LOLoopbackFlow(t *testing.T) {
...
+			"fetch; 3lo loopback cached; refresh expired token",
+			[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json", "--refresh",
+				"--disableAutoOpenConsentPage",
+				"--consentPageInteractionTimeout", "30", "--consentPageInteractionTimeoutUnits", "seconds"},
+			"fetch-3lo-cached.golden",
+			false,
+		},
+	}
+
+	type LoopbackLogic struct {
```
Comment (andyrzhao) [inline integration/cli_test.go:418]: I think we need a high-level comment on why the loopback pre and post logic is needed, to explain port selection and loopback triggering etc. Also, the current logic has a process to "restore credentials file" - what if the test fails in the middle and the credentials file doesn't get restored? I feel like it may be better to create a new temp cred file ONCE at the start of the loopback tests and delete the file in a deferred block at the end of the loopback tests - so you don't have to look for an available port each test or do clean up of the file for each test - at that point, you may be able to just have a single logic step to trigger the loopback instead of pre/post. WDYT?

File: integration/cli_test.go [hunk 4]
```diff
@@ -303,7 +324,7 @@ func Test3LOFlow(t *testing.T) {
 	process3LOOutput := func(output string) string {
 		return removeCodeChallenge(output)
 	}
-	runTestScenariosWithInputAndProcessedOutput(t, tests, newFixture(t, "fake-verification-code.fixture").asFile(), process3LOOutput)
+	runTestScenariosWithInputAndProcessedOutput(t, tests, newFixture(t, "fake-verification-code.fixture").asFile(), process3LOOutput, nil, nil)
```
[the remark this answers is not in the document]
Comment (ulisesL) [reply integration/cli_test.go]: Thank you for the suggestion - It has been added.

File: integration/cli_test.go [hunk 5]
```diff
@@ -303,7 +324,7 @@ func Test3LOFlow(t *testing.T) {
...
 }
 
 // TODO: Enhance tests so that the entire loopback flow can be tested
```
[the remark this answers is not in the document]
Comment (ulisesL) [reply integration/cli_test.go]: Done - good catch!

File: integration/cli_test.go [hunk 6]
```diff
@@ -316,88 +337,166 @@ func Test3LOLoopbackFlow(t *testing.T) {
...
+			(*ll).cred = newFixture(t, path.Base(f))
+			(*ll).content = (*ll).cred.load()
+			re := regexp.MustCompile("\"http://localhost\"")
+			match := re.FindString((*ll).content)
+			newContent := strings.Replace((*ll).content, match, "\""+a+"\"", 1)
+			(*ll).cred.write(newContent)
+
+			// Triggering loopback logic
+			go func() {
+				for (*ll).quit != true {
```
[the remark this answers is not in the document]
Comment (ulisesL) [reply integration/cli_test.go:457]: Added, thanks!

File: integration/cli_test.go [hunk 7]
```diff
@@ -537,7 +636,21 @@ func TestServiceAccountImpersonationFlow(t *testing.T) {
 		return string(jsonString)
 	}
 
-	runTestScenariosWithInputAndProcessedOutput(t, tests, nil, processOutput)
+	runTestScenariosWithInputAndProcessedOutput(t, tests, nil, processOutput, nil, nil)
+}
+
+func getCredentialsFileName(tc *testCase) string {
```
[the remark this answers is not in the document]
Comment (ulisesL) [reply integration/cli_test.go:653]: I must have forgotten about this at some point. I have updated the comments. Thanks again.

File: integration/cli_test.go [hunk 8]
```diff
@@ -316,88 +337,166 @@ func Test3LOLoopbackFlow(t *testing.T) {
...
+			return
+		}
+
+		return preLogic, postLogic
 	}
 
+	pre, post := loopbackLogic()
+
 	process3LOOutput := func(output string) string {
 		re := regexp.MustCompile("redirect_uri=http%3A%2F%2Flocalhost%3A\\d+")
```
[the remark this answers is not in the document]
Comment (ulisesL) [reply integration/cli_test.go]: Thank you for the suggestion -done!

File: integration/fixtures/fake-client-secrets-3lo-loopback.json (hunk 1 above)
Comment (ulisesL) [reply integration/fixtures/fake-client-secrets-3lo-loopback.json]: Introduced new email addresses as previously discussed.

File: integration/cli_test.go (hunk 3 above)
Comment (ulisesL) [reply integration/cli_test.go]: Hi Andy, 

the following has been added:
- temp file is created and deleted per test case
- added logic to prevent looping forever when trying to reach for the code and state endpoint handler
- port is only searched for once and reused during each test case

the following was not changed:
- 2 step process: pre/post logic

NOTE: post helps on putting an end to the loopback logic in case the command execution ends unexpectedly/prematurely.

I believe this is a good compromise between your proposal and the original implementation. However, I am open to new suggestions.

File: integration/cli_test.go [hunk 9]
```diff
@@ -112,12 +128,75 @@ func runTestScenariosWithInput(t *testing.T, tests []testCase, input *os.File) {
...
+			expected := golden.load()
+			if !reflect.DeepEqual(expected, actual) {
+				t.Fatalf("Expected: %v Actual: %v", expected, actual)
+			}
+		})
+	}
+}
+
 // Runs test cases where stdin input is needed and output needs to be processed before comparing to golden files.
 func runTestScenariosWithInputAndProcessedOutput(t *testing.T, tests []testCase, input *os.File, processOutput processOutput) {
```
Comment (andyrzhao) [inline integration/cli_test.go:196]: This function should call runTestScenariosWithAdvancedLogic and pass in nil nil as last two parameter, instead of duplicating the code. (See pattern in runTestScenariosWithInput)

File: integration/cli_test.go [hunk 10]
```diff
@@ -306,98 +395,184 @@ func Test3LOFlow(t *testing.T) {
...
+
+			// Modifiying credentials file temp copy: redirect uri
+			re := regexp.MustCompile(QUOTED_LOCALHOST)
+			match := re.FindString(fileContent)
+			newContent := strings.Replace(fileContent, match, "\""+loopbackAddr+"\"", 1)
+			(*ll).cred.write(newContent)
+
+			// Start loopback logic.
+			go func() {
+				timer := time.AfterFunc(CONSET_PAGE_DURATION, func() {
```
Comment (andyrzhao) [inline integration/cli_test.go:531]: Using timer here seems a bit overkill. I would convert the for loop condition to take an integer counter or something, so it ends after 30 loops max. Also, is this to safeguard against the scenario where postLogic doesn't get triggered? i.e. CMD crashes?

**Review by andyrzhao — CHANGES_REQUESTED** [review]
Looks good overall! Thanks for making the updates!

File: integration/cli_test.go [hunk 11]
```diff
@@ -306,98 +395,184 @@ func Test3LOFlow(t *testing.T) {
...
-// TODO: Once enhanced, uncomment and fix cache tests in this flow
 // TODO: Remove Test3LOFlow once the 3LO flow is deprecated
-// Test OAuth 3LO loopback flow with fake client secrets. Stops waiting for consent page interaction to advance the flow.
+// Test OAuth 3LO loopback flow with fake client secrets. It does not wait for consent page interaction.
+// Instead a post request with the code and state is sent to the loopback server to advance the flow.
 func Test3LOLoopbackFlow(t *testing.T) {
+
+	const (
+		// NOTE: Update all consent page settings accordingly if one is changed.
+		CONSET_PAGE_TIMEOUT       = "30"
```
[the remark this answers is not in the document]
Comment (ulisesL) [reply integration/cli_test.go]: Got it, thanks!

File: integration/cli_test.go [hunk 12]
```diff
@@ -92,6 +96,18 @@ func (tf *testFile) load() string {
 	return string(content)
 }
 
+// saveAs saves the currect instance into another file in the same directory.
+// The current instance is not substituded by the newly create one.
```
[the remark this answers is not in the document]
Comment (ulisesL) [reply integration/cli_test.go]: Thanks! It is fixed.

File: integration/cli_test.go (hunk 9 above)
Comment (ulisesL) [reply integration/cli_test.go:188]: Hi Andy, I had followed this pattern before. At some point, the two functions looks so different that I needed to diverge. I have restore the previous logic, which is exactly what you are looking for. Thanks for catching this horror :)

File: integration/cli_test.go [hunk 13]
```diff
@@ -112,12 +128,75 @@ func runTestScenariosWithInput(t *testing.T, tests []testCase, input *os.File) {
...
+type postCommandLogic func(loopbackAddr string)
+
+// Runs tests where extra logic needs to be added before/after the command execution.
+//
+// input processOutput: processes test output before comparing to golden files
+// input preCmdLogic: additional logic excuted once per test case before command excution. It allows for test argument manipulation.
+// input postCmdLogic: additional logic excuted once per test case after command excution. It allows for cleanup.
+func runTestScenariosWithAdvancedLogic(t *testing.T, tests []testCase, input *os.File, processOutput processOutput,
+	preCmdLogic preCommandLogic, postCmdLogic postCommandLogic) {
+	// Looking for available port.
```
[the remark this answers is not in the document]
Comment (ulisesL) [reply integration/cli_test.go]: Moved the logic inside loopbackLogic, which makes a little more sense - WDYT?

File: integration/cli_test.go (hunk 10 above)
Comment (ulisesL) [reply integration/cli_test.go]: Thanks for your suggestion and please excuse me if I am over explaining. 

Yes, it does seem a little overkilled, but I do prefer it over a max try counter. Imagine that we are running the tests in a very slow machine. It takes 40 seconds to start serving the POST request and so all tests are failing for this flow. To fix it all is needed is to change the CONSENT_[CONSTANTS] - all of them. If a counter is used, then the SWE will need to dive into more complicated code to change the max try.  

In theory, the safety-net should never get trigger. As you pointed out, it should only get triggered when the postLogic is not configured correctly or when a crash occurs. If you believe I am being too cautious we can remove the timeout in its entirety. So there will be no need for a max counter either. Just  a friendly reminder: the loop is running async.  

Thanks!

**Review by andyrzhao — APPROVED** [review]
LGTM!

File: integration/cli_test.go (hunk 10 above)
Comment (andyrzhao) [reply integration/cli_test.go]: Sg, we can keep it as is for now. I general I don't think we need to be as cautious with test code since we are running it in a controlled environment, so if it passes once, it should always pass, and not trigger rare edge cases - and the consequence of a test failing on rare edge cases is also low. For example, the current tests before loopback support was added already has some shortcuts such as using port 8080 for mock server, and we have not ran into port collision problem yet, but we should probably fix that at some point too.
